File: backend/backend/src/main/java/com/complementasenac/backend/controller/CoordenadorController.java
package com.complementasenac.backend.controller;

import com.complementasenac.backend.model.AtividadeCoordenadorModel;
import com.complementasenac.backend.model.CoordenadorPerfilModel;
import com.complementasenac.backend.model.CoordenadorResumoModel;
import com.complementasenac.backend.model.AlunoAdminModel;
import com.complementasenac.backend.service.AdminAlunoService;
import com.complementasenac.backend.service.CoordenadorService;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/coordenador")
public class CoordenadorController {

    private final CoordenadorService coordenadorService;
    private final AdminAlunoService adminAlunoService;

    public CoordenadorController(CoordenadorService coordenadorService, AdminAlunoService adminAlunoService) {
        this.coordenadorService = coordenadorService;
        this.adminAlunoService = adminAlunoService;
    }

    @GetMapping("/atividades")
    public List<AtividadeCoordenadorModel> listarPendentes() {
        return coordenadorService.listarPendentes();
    }

    @GetMapping("/atividades/todas")
    public List<AtividadeCoordenadorModel> listarTodas() {
        return coordenadorService.listarTodas();
    }

    @PostMapping("/atividades/{id}/decisao")
    public ResponseEntity<?> decidirAtividade(@PathVariable Long id, @RequestBody Map<String, String> payload) {
        String status = payload.getOrDefault("status", "PENDENTE").toUpperCase();
        return coordenadorService.decidir(id, status)
                .<ResponseEntity<?>>map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    @GetMapping("/resumo")
    public CoordenadorResumoModel resumo() {
        return coordenadorService.resumo();
    }

    @GetMapping("/perfil")
    public CoordenadorPerfilModel perfil(HttpServletRequest request) {
        String uid = (String) request.getAttribute("uid");
        String email = (String) request.getAttribute("email");
        if (uid == null || uid.isBlank()) {
            uid = "uid-coordenador-local";
        }
        if (email == null || email.isBlank()) {
            email = "coordenador@senac.pe.br";
        }
        return coordenadorService.perfil(uid, email);
    }

    @GetMapping("/alunos")
    public List<AlunoAdminModel> listarAlunos() {
        return adminAlunoService.listar();
    }
}

File: backend/app/controllers/api/coordenador_controller.rb
# frozen_string_literal: true

module Api
  class CoordenadorController < ApplicationController
    include FirestoreApiBase

    before_action :require_usuario_doc!, only: %i[perfil]

    def atividades
      render_camel_json(coordenador_service.listar_pendentes)
    end

    def atividades_todas
      render_camel_json(coordenador_service.listar_todas)
    end

    def decisao
      payload = payload_hash
      status = payload[:status].to_s.upcase.presence || "PENDENTE"
      horas = payload[:horasAprovadas]
      horas_aprovadas = horas.present? ? horas.to_i : nil
      justificativa = payload[:justificativa]

      result = coordenador_service.decidir(
        params[:id],
        status,
        horas_aprovadas: horas_aprovadas,
        justificativa: justificativa
      )

      if result
        render_camel_json(result)
      else
        head :not_found
      end
    end

    def resumo
      render_camel_json(coordenador_service.resumo)
    end

    def perfil
      render_camel_json(
        coordenador_service.perfil(current_uid, current_email, usuario_doc: current_usuario_doc)
      )
    end

    def alunos
      render_camel_json(admin_aluno_service.listar)
    end
  end
end
